src/PortAudio.jl

```diff
@@ -91,83 +93,50 @@ struct PortAudioStream{T}
-        outchans = outchans == -1 ? outdev.maxoutchans : outchans
-        inparams = if (inchans == 0)
-            Ptr{Pa_StreamParameters}(0)
-        else
-            Ref(Pa_StreamParameters(indev.idx, inchans, type_to_fmt[T], latency, C_NULL))
-        end
-        outparams = if (outchans == 0)
-            Ptr{Pa_StreamParameters}(0)
-        else
-            Ref(Pa_StreamParameters(outdev.idx, outchans, type_to_fmt[T], latency, C_NULL))
-        end
-        # finalizer(close, this)
-        pointer_ref = @stderr_as_debug Pa_OpenStream(
-            inparams,
-            outparams,
-            sr,
-            0,
-            paNoFlag,
-            nothing,
-            nothing,
-        )
-        sink_buffer = Buffer{T}(outdev, outchans)
-        source_buffer = Buffer{T}(indev, inchans)
-        Pa_StartStream(pointer_ref[])
-        this = new(
-            sr, 
-            latency, 
-            pointer_ref, 
-            warn_xruns,
-            recover_xruns,
-            sink_buffer,
-            source_buffer
+function make_parameters(device, channels, T, latency, host_api_specific_stream_info)
+    if channels == 0
+        Ptr{Pa_StreamParameters}(0)
+    else
+        Ref(
+            Pa_StreamParameters(
+                device.idx,
+                channels,
+                type_to_fmt[T],
+                latency,
+                convert_nothing(host_api_specific_stream_info),
+            ),
         )
-        # pre-fill the output stream so we're less likely to underrun
-        prefill_output(this.sink)
+    end
+end
 
-        this
+function fill_max_channels(channels, bounds)
+    if channels === max
+        bounds.max_channels
+    else
+        channels
     end
 end
 
 function recover_xrun(stream::PortAudioStream)
-    playback = nchannels(stream.sink) > 0
-    capture = nchannels(stream.source) > 0
-    if playback && capture
+    sink = stream.sink
+    source = stream.source
+    if nchannels(sink) > 0 && nchannels(source) > 0
         # the best we can do to avoid further xruns is to fill the playback buffer and
         # discard the capture buffer. Really there's a fundamental problem with our
         # read/write-based API where you don't know whether we're currently in a state
         # when the reads and writes should be balanced. In the future we should probably
         # move to some kind of transaction API that forces them to be balanced, and also
         # gives a way for the application to signal that the same number of samples
         # should have been read as written.
-        discard_input(stream.source)
-        prefill_output(stream.sink)
+        discard_input(source)
+        prefill_output(sink)
     end
 end
 
-function defaultlatency(devices...)
-    maximum(d -> max(d.highoutputlatency, d.highinputlatency), devices)
+function defaultlatency(input_device, output_device)
+    max(input_device.input_bounds.high_latency, output_device.output_bounds.high_latency)
```

Comment:
this is a slightly smarter way to go about things. output bounds shouldn't matter for the input device, and vice versa.